File: client/project/src/main/java/org/bsc/dcc/vcv/AnalyticsRecorder.java
package org.bsc.dcc.vcv;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.io.*;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;


public class AnalyticsRecorder {

	
	private static final Logger logger = LogManager.getLogger("AllLog");
	protected final String workDir;
	protected final String folderName;
	protected final String experimentName;
	protected final String system;
	protected final String test;
	protected final int instance;
	private final BufferedWriter writer;
	
	
	public AnalyticsRecorder(String workDir, String folderName, String experimentName,
			String system, String test, int instance) {
		this.workDir = workDir;
		this.folderName = folderName;
		this.experimentName = experimentName;
		this.system = system;
		this.test = test;
		this.instance = instance;
		BufferedWriter writerTemp;
		try {
			File logFile = new File(this.workDir + "/" + this.folderName + "/analytics/" + 
						this.experimentName + "/" + this.test + "/" + this.instance + "/analytics.log");
			logFile.getParentFile().mkdirs();
			writerTemp = new BufferedWriter(new FileWriter(logFile));
		}
		catch (Exception e) {
			writerTemp = null;
			e.printStackTrace();
			this.logger.error(e);
		}
		this.writer = writerTemp;
	}
	
	
	protected void message(String msg) {
		//logger.info(msg);
		try {
			this.writer.write(msg);
			this.writer.newLine();
			this.writer.flush();
		}
		catch (Exception e) {
			e.printStackTrace();
			this.logger.error("Error in AnalyticsRecorder message.");
			this.logger.error(e);
			this.logger.error(AppUtil.stringifyStackTrace(e));
		}
	}
	
	
	public void header() {
		String[] titles = {"QUERY", "RUN", "SUCCESSFUL", "DURATION", "RESULTS_SIZE", "SYSTEM", 
						   "STARTDATE_EPOCH", "STOPDATE_EPOCH", "DURATION_MS", "STARTDATE",
						   "STOPDATE", "TUPLES"};
		StringBuilder builder = new StringBuilder();
		for(int i = 0; i < titles.length - 1; i++)
			builder.append(String.format("%-25s|", titles[i]));
		builder.append(String.format("%-25s", titles[titles.length-1]));
		if( this.system.equals("bigquery") )
			builder.append(String.format("|%-25s", "BYTES_BILLED"));
		this.message(builder.toString());
	}
	
	
	public void record(QueryRecord queryRecord) {
		int spaces = 25;
		String colFormat = "%-" + spaces + "s|";
		StringBuilder builder = new StringBuilder();
		builder.append(String.format(colFormat, queryRecord.getQuery()));
		builder.append(String.format(colFormat, queryRecord.getRun()));
		builder.append(String.format(colFormat, queryRecord.isSuccessful()));
		long durationMs = queryRecord.getEndTime() - queryRecord.getStartTime();
		String durationFormatted = String.format("%.3f", ((double) durationMs / 1000.0d));
		builder.append(String.format(colFormat, durationFormatted));
		builder.append(String.format(colFormat, queryRecord.getResultsSize()));
		builder.append(String.format(colFormat, this.system));
		builder.append(String.format(colFormat, queryRecord.getStartTime()));
		builder.append(String.format(colFormat, queryRecord.getEndTime()));
		builder.append(String.format(colFormat, durationMs));
		Date startDate = new Date(queryRecord.getStartTime());
		String startDateFormatted = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS").format(startDate);
		builder.append(String.format(colFormat, startDateFormatted));
		Date endDate = new Date(queryRecord.getEndTime());
		String endDateFormatted = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS").format(endDate);
		builder.append(String.format(colFormat, endDateFormatted));
		builder.append(String.format("%-" + spaces + "s", queryRecord.getTuples()));
		if( queryRecord instanceof QueryRecordBigQuery)
			builder.append(String.format("%-" + spaces + "s", 
				((QueryRecordBigQuery)queryRecord).getBytesBilled()));
		this.message(builder.toString());
	}
	
	
	public void close() {
		try {
			this.writer.close();
		}
		catch (Exception e) {
			e.printStackTrace();
			this.logger.error("Error in AnalyticsRecorder close.");
			this.logger.error(e);
			this.logger.error(AppUtil.stringifyStackTrace(e));
		}
	}
	
	
}

File: scala_apps/tpcds-power-test/src/main/scala/AnalyticsRecorder.scala
import java.text.SimpleDateFormat
import java.util.Date
import java.io.BufferedWriter
import java.io.FileWriter
import java.io.File
import java.nio.file.Paths

class AnalyticsRecorder(val resultsDir: String,
                        val test: String,
                        val instance: Int,
                        val system: String) {

  var writer: BufferedWriter = null
  var logFile: File = null
  
  def createWriter(): Unit = {
    try {
      this.logFile = new File(
        Paths.get(this.resultsDir, this.test, "analytics",
          this.instance.toString, "analytics.csv").toString
      )
      logFile.getParentFile.mkdirs()
      this.writer = new BufferedWriter(new FileWriter(logFile))
    }
    catch {
       case e: Exception => {
        e.printStackTrace()
      }
    }
  } 

  def message(msg: String): Unit = {
   try {
        this.writer.write(msg)
        println(msg)
        this.writer.newLine()
        this.writer.flush()
    }
    catch {
      case e: Exception => {
        e.printStackTrace()
      }
    }
  }
    
  def header(): Unit = {
    val titles: Array[String] = Array("QUERY",
                                      "RUN",
                                      "SUCCESSFUL",
                                      "DURATION",
                                      "RESULTS_SIZE",
                                      "SYSTEM",
                                      "STARTDATE_EPOCH",
                                      "STOPDATE_EPOCH",
                                      "DURATION_MS",
                                      "STARTDATE",
                                      "STOPDATE",
                                      "TUPLES")
    val builder: StringBuilder = new StringBuilder()
    for (i <- 0 until titles.length - 1)
      builder.append(String.format("%-25s|", titles(i)))
    builder.append(String.format("%-25s", titles(titles.length - 1)))
    this.message(builder.toString)
  }
  
  
  def record(queryRecord: QueryRecord): Unit = {
    val spaces: Int = 25
    val colFormat: String = "%-" + spaces + "s|"
    val builder: StringBuilder = new StringBuilder()
    val formatQuery = String.format(colFormat, queryRecord.query.toString)
    builder.append(formatQuery)
    val formatRun = String.format(colFormat, queryRecord.run.toString)
    builder.append(formatRun)
    val formatSuccessful = String.format(colFormat, queryRecord.successful.toString)
    builder.append(formatSuccessful)
    val durationMs: Long = queryRecord.endTime - queryRecord.startTime
    val durationFormatted: String =
      String.format("%.3f", new java.lang.Double(durationMs / 1000.0d))
    val formatDuration = String.format(colFormat, durationFormatted.toString)
    builder.append(formatDuration)
    val formatResults = String.format(colFormat, queryRecord.resultsSize.toString)
    builder.append(formatResults)
    val formatSystem = String.format(colFormat, this.system)
    builder.append(formatSystem)
    val formatStart = String.format(colFormat, queryRecord.startTime.toString)
    builder.append(formatStart)
    val formatEnd = String.format(colFormat, queryRecord.endTime.toString)
    builder.append(formatEnd)
    val formatDurationMs = String.format(colFormat, durationMs.toString)
    builder.append(formatDurationMs)
    val startDate: Date = new Date(queryRecord.startTime)
    val startDateFormatted: String =
      new SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS").format(startDate)
    builder.append(String.format(colFormat, startDateFormatted))
    val endDate: Date = new Date(queryRecord.endTime)
    val endDateFormatted: String =
      new SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS").format(endDate)
    val formatEndDate = String.format(colFormat, endDateFormatted.toString)
    builder.append(formatEndDate)
    val endColFormat: String = "%-" + spaces + "s"
    val formatTuples = String.format(endColFormat, queryRecord.tuples.toString)
    builder.append(formatTuples)
    this.message(builder.toString)
  }

  def close(): Unit = {
    try this.writer.close()
    catch {
      case e: Exception => {
        e.printStackTrace()
      }
    }
  }

  def getLogFileCanonicalPath(): String = {
    this.logFile.getCanonicalPath()
  }

  def getLogFileConstructorPath(): String = {
    this.logFile.getPath()
  }
  
}

